omnipaxos/src/ballot_leader_election.rs

```diff
@@ -225,15 +234,32 @@ impl BallotLeaderElection {
     }
 
     pub(crate) fn hb_timeout(&mut self) -> Option<Ballot> {
-        let my_connectivity = self.ballots.len() + 1;
+        // Update connectivity
+        self.ballots
+            .push((self.current_ballot, self.connectivity, self.happy));
+        let my_connectivity = self.ballots.len();
         self.connectivity = my_connectivity as Connectivity;
+
+        // Update happiness
+        let connected_to_qc_gossiped_leader = self
+            .ballots
+            .iter()
+            .filter(|(_, connectivity, _)| self.quorum.is_accept_quorum(*connectivity as usize))
+            .any(|(ballot, _, _)| ballot.pid == self.gossiped_leader.pid);
+        self.happy = connected_to_qc_gossiped_leader;
+
         let result: Option<Ballot> = if self.quorum.is_prepare_quorum(my_connectivity) {
             #[cfg(feature = "logging")]
             debug!(
                 self.logger,
                 "Received a majority of heartbeats, round: {}, {:?}", self.hb_round, self.ballots
             );
-            self.ballots.push((self.current_ballot, self.connectivity));
+            let unhappy_nodes = self.ballots.iter().filter(|&&(_, _, happy)| !happy).count();
+            if self.quorum.is_prepare_quorum(unhappy_nodes) {
+                self.current_ballot.n = self.gossiped_leader.n + 1;
+                self.gossiped_leader = self.current_ballot;
+                self.happy = true;
+            }
```

Comment:
Can we move this to check_leader()? I think it's easier to read when all the leader changing logic is in the same function.